### PR TITLE
Test to verify the hub restore when active hub down following failover and relocate of app

### PR DESCRIPTION
**Pull request by Akarsha-rai**
Test to verify the hub restore when active hub down following failover and relocate apps

### CODE REVIEW

File: ocs_ci/ocs/constants.py [hunk 1]
```diff
@@ -846,6 +846,7 @@
 MDR_BACKUP_SCHEDULE_YAML = os.path.join(
     TEMPLATE_MULTICLUSTER_DIR, "backupschedule.yaml"
 )
+MDR_RESTORE_YAML = os.path.join(TEMPLATE_MULTICLUSTER_DIR, "restore.yaml")
```
Comment (am-agrawa) [inline ocs_ci/ocs/constants.py:849]: This restore yaml is common for both RDR and MDR, hence a generic name would make it flexible for re-using.

File: ocs_ci/helpers/dr_helpers.py [hunk 2]
```diff
@@ -860,3 +869,42 @@ def gracefully_reboot_ocp_nodes(
...
+        bool: True if the status is in succeed state, else False
+        switch_ctx (int): The cluster index by the cluster name
+
+    """
+
+    restore_index = config.cur_index
+    config.switch_ctx(switch_ctx) if switch_ctx else config.switch_acm_ctx()
+    drpolicy_obj = ocp.OCP(kind=constants.DRPOLICY)
+    status = drpolicy_obj.get().get("items")[0].get("status").get("conditions")[0]
+    if status.get("reason") == "Succeeded":
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers.py:905]: Some sort of re-try and wait mechanism would be good as DRPolicy takes 5-10mins to get validated and we would not want to fail the test immediately if the condition is not met.

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+from ocs_ci.utility import version
+from ocs_ci.utility.utils import TimeoutSampler
+
+
+logger = logging.getLogger(__name__)
+
+
+@tier4a
+@turquoise_squad
+class TestApplicationFailoverAndRelocateWhenZoneDown:
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:48]: file name should be based upon the class/function name IMO for better readibility.

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    """
+    Test failover and relocate all apps in a single zone after a zone disruption
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:50]: Is it the complete zone disruption? The zone might also have any of the managed clusters running.

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    @pytest.mark.polarion_id("OCS-XXXX")
+    def test_application_failover_and_relocate(
+        self,
+        setup_acm_ui,
+        nodes_multicluster,
+        dr_workload,
+    ):
+
+        """
+        Tests to verify failover and relocate all apps in a single zone after a zone disruption
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:85]: As above

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+        """
+
+        if config.RUN.get("mdr_failover_via_ui"):
+            ocs_version = version.get_semantic_ocs_version_from_config()
+            if ocs_version <= version.VERSION_4_12:
+                logger.error(
+                    "ODF/ACM version isn't supported for Failover/Relocate operation"
+                )
+                raise NotImplementedError
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:95]: We may no longer need this due to release-branches. Kindly check.

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        # Create backup-schedule on active hub
+        create_backup_schedule()
+        # ToDo: To verfiy all the backups are taken
+        wait_time = 300
+        logger.info(f"Wait {wait_time} until backup is taken ")
+        time.sleep(wait_time)
+
+        # Install gitops operator on passive hub
+        dep_obj = Deployment()
+        dep_obj.deploy_gitops_operator(switch_ctx=get_passive_acm_index())
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:120]: Am I missing the reference for steps to deploy MCO on passive hub?

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        self.primary_cluster_name = get_current_primary_cluster_name(
+            namespace=workload.workload_namespace
+        )
+        secondary_cluster_name = get_current_secondary_cluster_name(
+            workload.workload_namespace
+        )
+
+        # Create backup-schedule on active hub
+        create_backup_schedule()
+        # ToDo: To verfiy all the backups are taken
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:113]: This verification might be important for the same PR

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+        # Shutdown active hub nodes
+        logger.info("Shutting down all the nodes of active hub")
+        nodes_multicluster[active_hub_index].stop_nodes(active_hub_cluster_node_objs)
+        logger.info(
+            "All nodes of active hub zone are powered off, "
+            f"wait {wait_time} seconds before restoring in passive hub"
+        )
+
+        # Restore new hub
+        restore_backup()
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:136]: Did we switch the context to passive hub before restoring the backup?

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+        logger.info(f"Wait {wait_time} until restores are taken ")
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py:137]: Post restore, status of Velero might be important as a verification step

File: ocs_ci/deployment/deployment.py [hunk 11]
```diff
@@ -217,6 +217,52 @@ def do_deploy_submariner(self):
             submariner = Submariner()
             submariner.deploy()
 
+    def deploy_gitops_operator(self, switch_ctx=None):
```
Comment (hnallurv) [inline ocs_ci/deployment/deployment.py:220]: As discussed earlier, please make sure to cover appset based apps also.

Comment (Akarsha-rai) [reply ocs_ci/deployment/deployment.py:223]: added

File: ocs_ci/helpers/dr_helpers.py (hunk 2 above)
Comment (Akarsha-rai) [reply ocs_ci/helpers/dr_helpers.py:1051]: added retry method

File: ocs_ci/ocs/constants.py (hunk 1 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/constants.py]: changed to DR_RESTORE_YAML

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py (hunk 4 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py]: That was by mistake, changed now

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py (hunk 5 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py]: That was by mistake, changed now

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py (hunk 6 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py]: I'm not sure, it didn't work removing this so kept as it is

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py (hunk 7 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py]: MCO is installed when deploying cluster, only gitops and backupschedule yaml needs be handled in test case

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py (hunk 8 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py]: Agreed, but to verify all the backups are taken we need exact/accurate steps and also it is lengthy one, so wil take in next PR

File: tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py (hunk 10 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_active_hub_down_and_restore.py]: Agreed, and added verification to check is restore succeeded

**Review by am-agrawa — APPROVED** [review]

**Review by prsurve — APPROVED** [review]

File: tests/functional/disaster-recovery/metro-dr/test_active_hub_down_and_restore.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+    @pytest.mark.polarion_id("OCS-XXXX")
+    def test_application_failover_and_relocate(
+        self, setup_acm_ui, nodes_multicluster, dr_workload
+    ):
+
+        """
+        Tests to verify failover and relocate all appswhen active hub down and restored
+
+        """
+
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/functional/disaster-recovery/metro-dr/test_active_hub_down_and_restore.py:86]: This is handled as a part of the deployment

[the remark this answers is not in the document]
Comment (prsurve) [reply tests/functional/disaster-recovery/metro-dr/test_active_hub_down_and_restore.py:86]: do you think we should have any requirement check before running the test ?

[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/functional/disaster-recovery/metro-dr/test_active_hub_down_and_restore.py:86]: make sense, we can add a step to verify `BackupStorageLocation` is in `Available` or not, only if it is Available proceed to test

**Review by parikshithb — APPROVED** [review]

File: tests/functional/disaster-recovery/metro-dr/test_active_hub_down_and_restore.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (prsurve) [reply tests/functional/disaster-recovery/metro-dr/test_active_hub_down_and_restore.py:86]: ack you can take it in new pr

**Review by prsurve — APPROVED** [review]

**Review by vavuthu — APPROVED** [review]

**Pull request merged**